kratos/geometries/nurbs_shape_function_utilities/nurbs_curve_shape_functions.h

```diff
@@ -0,0 +1,402 @@
+        const Vector& rKnots,
+        const IndexType Span,
+        const Vector& rWeights,
+        const double ParameterT)
+    {
+        // compute B-Spline shape
+        ComputeBSplineShapeFunctionValuesAtSpan(rKnots, Span, ParameterT);
+
+        // apply weights
+        std::vector<double> weightedSums(NumberOfShapeFunctionRows());
+
+        for (IndexType k = 0; k < NumberOfShapeFunctionRows(); k++) {
+            weightedSums[k] = 0;
+
+            for (IndexType i = 0; i < NumberOfNonzeroControlPoints(); i++) {
+                const IndexType poleIndex = GetFirstNonzeroControlPoint() + i;
+                ShapeFunctionValue(i, k) *= rWeights(poleIndex);
+                weightedSums[k] += ShapeFunctionValue(i, k);
+            }
+        }
+
+        for (IndexType k = 0; k < NumberOfShapeFunctionRows(); k++) {
+            for (IndexType i = 1; i <= k; i++) {
+                const double a = NurbsUtilities::GetBinomCoefficient(k, i) *
+                    weightedSums[i];
+
+                for (IndexType p = 0; p < NumberOfNonzeroControlPoints(); p++) {
+                    ShapeFunctionValue(p, k) -=
+                        a * ShapeFunctionValue(p, k - i);
+                }
+            }
+
+            for (IndexType p = 0; p < NumberOfNonzeroControlPoints(); p++) {
+                ShapeFunctionValue(p, k) /= weightedSums[0];
+            }
+        }
+    }
+
+    ///@}
+
+private:
+    ///@name Private Operations
+    ///@{
+
+    /*
+    * @brief Provides the shape function value depending to the derivative row and
+    *        the index of the control point.
+    *        For curves the indices for derivatives are as following:
+    *        0-> N | 1-> dN/de | 2-> ddN/dde | ...
+    */
+    double& ShapeFunctionValue(const IndexType NonzeroControlPointIndex, const IndexType DerivativeRow)
+    {
+        const IndexType index = NurbsUtilities::GetVectorIndexFromMatrixIndices(
+            NumberOfNonzeroControlPoints(), NumberOfShapeFunctionRows(),
+            NonzeroControlPointIndex, DerivativeRow);
+
+        return mValues[index];
+    }
+
+    double& Ndu(const IndexType DerivativeRow, const IndexType NonzeroControlPointIndex)
+    {
+        const IndexType index = NurbsUtilities::GetVectorIndexFromMatrixIndices(
+            NumberOfNonzeroControlPoints(), NumberOfShapeFunctionRows(), NonzeroControlPointIndex, DerivativeRow);
+
+        return mNdu[index];
+    }
+
+    /*
+    * @brief resets the shape function container.
+    */
+    void ClearValues()
+    {
+        const SizeType number_of_values = NumberOfNonzeroControlPoints() * NumberOfShapeFunctionRows();
+
+        mValues = ZeroVector(number_of_values);
+    }
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+
+    SizeType mPolynomialDegree;
+    SizeType mDerivativeOrder;
+    Vector mValues;
+    Vector mLeft;
+    Vector mRight;
+    Vector mNdu;
+    Vector mA;
+    Vector mB;
+    IndexType mFirstNonzeroControlPoint;
```

Comment:
please add descriptions since the names are not descriptive

Reply:
actually, it is hard to add a documentation here. Those values are operation and temp variables for the computation of the NURBS. Those are used as they are called in the NURBS-bool.

Reply:
completely agree with @tteschemacher. Internal variables + same naming as in the nurbs book.

Reply:
could you then please add this reference? :)
Makes it much easier for others to follow

Reply:
I think @tteschemacher has already added several references (_From Piegl and Tiller, The NURBS Book, Algorithm ..._)